### PR TITLE
Better enclosures

### LINKED ISSUE #1668: Duplicate images
dup images from https://www.yaplakal.com/news.xml

### LINKED ISSUE #4702: [Feature] Enclosures as part of greader API
**Is your feature request related to a problem? Please describe.**
It would be nice to easily access enclosures like thumbnails, podcast audio tracks etc. from client applications without parsing the html content.

**Describe the solution you’d like**
Extend the google reader API entries with enclosures like most other APIs do. Examples would be the [feedly API](https://developer.feedly.com/v3/entries/), [Nextcloud News API](https://github.com/nextcloud/news/blob/master/docs/api/api-v1-3.md#items), [tt-rss API (called attachments)](https://tt-rss.org/wiki/ApiReference) and [feedbin API](https://github.com/feedbin/feedbin-api/blob/master/content/entries.md).

**Describe alternatives you’ve considered**
It seems like enclosures are somehow embedded in the article html content:
```
<div class=\"enclosure\"><p class=\"enclosure-title\">💧🔥 BEST COMBO COMMANDER | Jhoira, Ageless Innovator | Competitive Historic Brawl [MTG Arena]</p><p><img class=\"enclosure-thumbnail\" src=\"https://i4.ytimg.com/vi/_qUe1nd09Ic/hqdefault.jpg\" alt=\"\" /></p><p class=\"enclosure-content\">
```

So in theory one could parse the content and extract the data from there. But that would mean a lot of special handling for FreshRSS compared to other popular feed services in the client app.

**Additional context**
Issue filed against NewsFlash client: https://gitlab.com/news-flash/news_flash_gtk/-/issues/374

The google reader API has already been extended by others with minor new features (e.g. friends lists https://github.com/theoldreader/api#friends).

### LINKED ISSUE #4974: [BUG] Enclosure shows only 1 image of many
**Describe the bug**
Example feed: https://raw.githubusercontent.com/math-GH/testfeed/main/test-feed.xml

"News 4" has 2 images
grafik

but only 1 image is shown
grafik

**Environment information (please complete the following information):**
- FreshRSS version: 1.20.0-dev

### PR DESCRIPTION
**Pull request by Alkarex**
#fix https://github.com/FreshRSS/FreshRSS/issues/4702
Improvement of https://github.com/FreshRSS/FreshRSS/pull/2898

### CODE REVIEW

**Comment by math-GH** [comment]
@Alkarex am I allowed to add my changes, that moves the HTML from `content()` to the `.phtml`? Or should I wait till this PR is merged and will open an additional PR later?

**Comment by Alkarex** [comment]
> @Alkarex am I allowed to add my changes, that moves the HTML from `content()` to the `.phtml`? Or should I wait till this PR is merged and will open an additional PR later?

Do you mean https://github.com/FreshRSS/FreshRSS/pull/4943 ?

**Comment by Alkarex** [comment]
I have just spotted a couple of bugs in SimplePie that needs fixing for this PR to work with various encodings

**Comment by math-GH** [comment]
> 



> > @Alkarex am I allowed to add my changes, that moves the HTML from `content()` to the `.phtml`? Or should I wait till this PR is merged and will open an additional PR later?
> 
> Do you mean #4943 ?

#4943 was just the proof of concept.

> I have just spotted a couple of bugs in SimplePie that needs fixing for this PR to work with various encodings

ok, then it will make sense to wait till this PR is done.

File: lib/SimplePie/SimplePie/Enclosure.php [hunk 1]
```diff
@@ -627,7 +627,7 @@ public function get_link()
 	{
 		if ($this->link !== null)
 		{
-			return urldecode($this->link);
+			return $this->link;
```
Comment (Alkarex) [inline lib/SimplePie/SimplePie/Enclosure.php:630]: This line was introduced in https://github.com/simplepie/simplepie/commit/8a5f6013c4405cc8d5ccb27f2d76c78e8657df53#diff-a710c236cc7775672edbe5840d8fb96b0e91b0d56eff709f6eace2da47c1fe43R349-R359

Seems to be a similar error than https://github.com/simplepie/simplepie/commit/efb1d8e34c39276fb88fff0dcff01c673bec7e18

File: lib/SimplePie/SimplePie/Item.php [hunk 2]
```diff
@@ -427,7 +427,16 @@ public function get_thumbnail()
...
 			{
-				$this->data['thumbnail'] = $return[0]['attribs'][''];
+				$thumbnail = $return[0]['attribs'][''];
+				if (empty($thumbnail['url']))
+				{
+					$this->data['thumbnail'] = null;
+				}
+				else
+				{
+					$thumbnail['url'] = $this->sanitize($thumbnail['url'], SIMPLEPIE_CONSTRUCT_IRI);
```
Comment (Alkarex) [inline lib/SimplePie/SimplePie/Item.php]: The thumbnail URL was previously returned unsanitized and unescaped, as opposed to other URLs returned by SimplePie

**Comment by Alkarex** [comment]
@math-GH Tests very much welcome already.

I need to look into another SimplePie potential issue with a lack of `absolutize_url` for the thumbnail URL, but I will continue a bit later

File: lib/SimplePie/SimplePie/Item.php (hunk 2 above)
Comment (Alkarex) [reply lib/SimplePie/SimplePie/Item.php]: TODO: provide a `base` as parameter for `absolutize_url` to work. Check that other enclosure URLs are correct in this regard.

**Comment by math-GH** [comment]
@Alkarex I added 2 issues regarding enclosures

File: lib/SimplePie/SimplePie/Item.php [hunk 3]
```diff
@@ -2847,9 +2856,9 @@ public function get_enclosures()
 				}
 			}
 
-			if ($enclosure = $this->get_item_tags(SIMPLEPIE_NAMESPACE_RSS_20, 'enclosure'))
+			foreach ((array) $this->get_item_tags(SIMPLEPIE_NAMESPACE_RSS_20, 'enclosure') as $enclosure)
```
Comment (Alkarex) [inline lib/SimplePie/SimplePie/Item.php]: Fix https://github.com/FreshRSS/FreshRSS/issues/4974

Comment (math-GH) [reply lib/SimplePie/SimplePie/Item.php]: can confirm, that it fix it.

Comment (math-GH) [reply lib/SimplePie/SimplePie/Item.php]: It breaks the article thumbnails.
Before: the article thumbnails took the first image of article or enclosure.
After: no thumbnail generated from enclosure.

File: lib/SimplePie/SimplePie/Enclosure.php (hunk 1 above)
Comment (Alkarex) [reply lib/SimplePie/SimplePie/Enclosure.php:630]: Upstream PR https://github.com/simplepie/simplepie/pull/768

File: lib/SimplePie/SimplePie/Item.php (hunk 3 above)
Comment (Alkarex) [reply lib/SimplePie/SimplePie/Item.php]: Upstream PR https://github.com/simplepie/simplepie/pull/769

File: lib/SimplePie/SimplePie/Item.php (hunk 2 above)
Comment (Alkarex) [reply lib/SimplePie/SimplePie/Item.php]: Upstream PR: https://github.com/simplepie/simplepie/pull/770

File: lib/SimplePie/SimplePie/Item.php (hunk 3 above)
Comment (Alkarex) [reply lib/SimplePie/SimplePie/Item.php]: > It breaks the article thumbnails. Before: the article thumbnails took the first image of article or enclosure. After: no thumbnail generated from enclosure.

@math-GH Please try again ( https://github.com/FreshRSS/FreshRSS/pull/4944/commits/c5c09ea9547faeccd5d44258cb52a2fc2a3cf4ce )

Comment (math-GH) [reply lib/SimplePie/SimplePie/Item.php]: Nice. It works fine!

**Comment by math-GH** [comment]
This PR breaks the multi enclosures of Atom feeds.

Example feed: https://github.com/math-GH/testfeed/blob/main/Atomfeed-with-enclosure.xml
This 1 article has 4 enclosure images.

`main` branch: all 4 images will be shown
this PR: only 1 image will be shown

**Comment by Alkarex** [comment]
> This PR breaks the multi enclosures of Atom feeds.
> 
> Example feed: https://github.com/math-GH/testfeed/blob/main/Atomfeed-with-enclosure.xml This 1 article has 4 enclosure images.
> 
> `main` branch: all 4 images will be shown this PR: only 1 image will be shown

@math-GH This is because the HTML content already contains those images, in order to avoid duplicates (cf. https://github.com/FreshRSS/FreshRSS/issues/1668 )

https://github.com/FreshRSS/FreshRSS/pull/4944/files#diff-318f03d8ac66816058c42f26b8ed4c5ff71306b6250d7096ad25e0a15813e8efR125-R127

The four enclosures are correctly stored in database and provided again e.g. via API or RSS view

```json
{
	"enclosures": [{
			"url": "https://0.gravatar.com/avatar/347dfeae267d8d9b7c8bf959b3b97baa?s=96&amp;d=identicon&amp;r=G",
			"title": "tomsgedankenblog",
			"medium": "image"
		}, {
			"url": "https://agileverwaltungorg.files.wordpress.com/2022/10/grafik-4.png?w=605",
			"medium": "image"
		}, {
			"url": "https://agileverwaltungorg.files.wordpress.com/2022/10/grafik-3.png?w=605",
			"medium": "image"
		}, {
			"url": "https://agileverwaltungorg.files.wordpress.com/2022/10/grafik-5.png?w=605",
			"medium": "image"
		}
	]
}
```

**Comment by math-GH** [comment]
@Alkarex  That is a great improvement ❤️

**Comment by Alkarex** [comment]
Let's give it a try 🚀

**Comment by math-GH** [comment]
> > This PR breaks the multi enclosures of Atom feeds.
> > Example feed: https://github.com/math-GH/testfeed/blob/main/Atomfeed-with-enclosure.xml This 1 article has 4 enclosure images.
> > `main` branch: all 4 images will be shown this PR: only 1 image will be shown
> 
> @math-GH This is because the HTML content already contains those images, in order to avoid duplicates (cf. #1668 )
> 
> ```

I do not know if it is a bug or improvement or a feature:
The thumbnail is now the blue avatar, that is the first enclosure.
What would make more sense? 
To take the first enclosure as thumbnail? Or take the first image in the article?

**Comment by Alkarex** [comment]
> I do not know if it is a bug or improvement or a feature: The thumbnail is now the blue avatar, that is the first enclosure. What would make more sense? To take the first enclosure as thumbnail? Or take the first image in the article?

@math-GH Compared to before, the sole difference should be that we now support the dedicated `thumbnail` RSS element, which we then use as first priority. Before this PR, we were already searching for the first image in enclosures and then in the HTML body.

**Pull request merged**